libs/execution_context/ebpf_program.c

```diff
@@ -437,29 +438,39 @@ ebpf_program_load_byte_code(ebpf_program_t* program, ebpf_instruction_t* instruc
 void
 ebpf_program_invoke(_In_ const ebpf_program_t* program, _In_ void* context, _Out_ uint32_t* result)
 {
+    uint64_t count = 0;
+    const ebpf_program_t* current_program = program;
     if (!program || program->program_invalidated) {
         *result = 0;
         return;
     }
 
-    if (program->parameters.code_type == EBPF_CODE_NATIVE) {
-        ebpf_program_entry_point_t function_pointer;
-        function_pointer = (ebpf_program_entry_point_t)(program->code_or_vm.code.code_pointer);
-        *result = (function_pointer)(context);
-    } else {
-        uint64_t out_value;
-        int ret = (uint32_t)(ubpf_exec(program->code_or_vm.vm, context, 1024, &out_value));
-        if (ret < 0) {
-            *result = ret;
+    for (count = 0; count < EBPF_MAX_TAIL_CALL_DEPTH; count++) {
+        uint64_t unwind_context = 0;
+        if (current_program->parameters.code_type == EBPF_CODE_NATIVE) {
+            ebpf_program_entry_point_t function_pointer;
+            function_pointer = (ebpf_program_entry_point_t)(current_program->code_or_vm.code.code_pointer);
+            *result = (function_pointer)(context, 1024, &unwind_context);
```

Comment:
These 2 hidden parameters will be passed using some registry. How will you ensure that no JIT compiled code uses/writes into these registry? Or are you going to make changes in the JIT compiler to always push them on stack and then make the tail call implementation aware of this hidden pointer variable?

I didnt see you modifying the tail call prototype BTW - would that be part of another change?

Reply:
Yes, the Windows x64 calling convention involves passing the first parameters via registers (RCX, RDX, R8, and R9). So, functions that don't need these registers ignore them. Functions that trigger unwind write to the address contained in R8 on unwinding.

JIT compiler was modified to always store the parameter registers on the stack at known locations so that JIT code can access them later.